### ISSUE TITLE
torch.nn.functional.nll_loss return unexpected nan

### ISSUE THREAD

**Issue by shizi626** [pytorch/pytorch#20406]
## 🐛 Bug

`torch.nn.functional.nll_loss` returns unexpected `nan` loss on some `ignore_index`  value
## To Reproduce

```python
import torch
import torch.nn.functional as F
```
The inputs and targets are following
```python
input = torch.tensor([[[-0.7104, -0.6762],
  [-0.7114, -0.6753],
  [-0.6960, -0.6903],
  [-0.6883, -0.6981],
  [-0.6932, -0.6930],
  [-0.6973, -0.6890],
  [-0.6973, -0.6890],
  [-0.6973, -0.6890],
  [-0.6973, -0.6890],
  [-0.6973, -0.6890],
  [-0.6973, -0.6890],
  [-0.6973, -0.6890]],
[[-0.6651, -0.7220],
  [-0.6567, -0.7310],
  [-0.6614, -0.7260],
  [-0.6499, -0.7384],
  [-0.6586, -0.7289],
  [-0.6550, -0.7328],
  [-0.6656, -0.7215],
  [-0.6512, -0.7370],
  [-0.6405, -0.7487],
  [-0.6484, -0.7400],
  [-0.6527, -0.7353],
  [-0.6593, -0.7282]]])
target=torch.tensor([[1,0,0,0,0,-1,-1,-1,-1,-1,-1,-1],
[0,0,0,0,0, 0, 1, 0, 0, 0, 0, 0]])


output = F.nll_loss(input.view(-1, input.size(-1)), target.contiguous().view(-1), ignore_index=-1, reduction='none')
```
this is the output result
```
output
tensor([0.6762, 0.7114, 0.6960, 0.6883, 0.6932, 0.0000, 0.0000, 0.0000, 0.0000,
           nan, 0.0000, 0.0000, 0.6651, 0.6567, 0.6614, 0.6499, 0.6586, 0.6550,
        0.7215, 0.6512, 0.6405, 0.6484, 0.6527, 0.6593])

```

## Expected behavior

I think the output result on `ignore_index` value (i.e. `-1` in this case) should be zero (or other values except `nan`), but I get `nan` on 10'th output result.
## Environment

```bash
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 384.130
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.4.2
[pip3] numpy==1.14.5
[pip3] torch==0.4.1
[pip3] torchvision==0.2.2.post3
[conda] Could not collect
```

BTW, the same bug occur at PyTorch version: 1.1.0

## Additional context

while on Win10 pytorch, the above example produces expected result:
```python
In[93]: output
Out[93]: 
tensor([6.7620e-01, 7.1140e-01, 6.9600e-01, 6.8830e-01, 6.9320e-01, 6.7210e-01,
        6.8620e-01, 6.8170e-01, 1.0842e-19, 1.9067e+00, 2.0692e+00, 1.6831e+00,
        6.6510e-01, 6.5670e-01, 6.6140e-01, 6.4990e-01, 6.5860e-01, 6.5500e-01,
        7.2150e-01, 6.5120e-01, 6.4050e-01, 6.4840e-01, 6.5270e-01, 6.5930e-01])
```

**Comment by soumith** [comment]
this seems to have been fixed with PyTorch nightly.

Atleast I cant reproduce it on Ubuntu 16.04 as the user said.

@shizi626 if this is not true with nightly version, can you reopen. You can install pytorch-nightly binaries via the install commands in the matrix at https://pytorch.org/ (select "Preview (nightly)  " option )
